README.md
# Archive Pages

Websites for your archives, served directly from your archive.org item.

## How it works

Step 1: Add an archive.org item

Step 2: Create a zip file of your website as a zip file with name archive-pages.zip

Step 3: Upload archive-pages.zip to your archive.org

Your archive pages website will be live at `https://<itemname>.onarchive.org/`

## Example

An archive of [Aaron Swartz's website](www.aaronsw.com) on hosted using Archive Pages.

<http://aaronsw-archive.onarchive.org/>

See the item on archive.org:
<https://archive.org/details/aaronsw-archive>

## How to run

```
$ docker-compose up -d 
```

Visit http://aaronsw-archive.local.onarchive.org:8080/

## License

Archive Pages is licensed under MIT License. Please see the LICENSE file for more details.


CHANGELOG.md
# Change Log

## Unreleased

## 0.3.0 -- 2020-09-06

* The sytem now expects a single zip file with name archive-paghes.zip to be present in the item
* The archive.yml is depricated

## 0.2.0 -- 2020-08-24

* Switched the implementation from Python to Go lang
* Added docker-compose to streamline development
* Added redis for caching metadata and Varnish for http caching
* Only the items that have archive.yml will be supported

## 0.1.0 -- 2020-07-27

* First working prototype of archive pages


ia.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"

	"gopkg.in/yaml.v2"
)

// Item represents an archive.org item
type Item struct {
	Identifier  string
	D1          string
	D2          string
	Dir         string
	LastUpdated int
	Root        string
}

// PagesMetadata provides the contents of the archive.yml file
type PagesMetadata struct {
	// Version  int    `yaml:"version"`
	ItemType string `yaml:"itemtype"`
	Root     string `yaml:"root"`
}

type ItemMetadata struct {
	Created     int        `json:"created"`
	D1          string     `json:"d1"`
	D2          string     `json:"d2"`
	Dir         string     `json:"dir"`
	Files       []ItemFile `json:"files"`
	LastUpdated int        `json:"item_last_updated"`
	Size        int        `json:"size"`
}

type ItemFile struct {
	Name  string `json:"name"`
	Mtime string `json:"mtime"`
	Size  string `json:"size"`
}

func GetItem(identifier string) (*Item, error) {
	var item *Item
	item = GetCacheItem(identifier)
	if item != nil {
		log.Printf("CACHE HIT: %s\n", identifier)
		return item, nil
	}
	log.Printf("CACHE MISS: %s\n", identifier)

	metadata, err := GetItemMetadata(identifier)
	if err != nil {
		return nil, err
	}

	var root string

	if hasArchivePagesZip(metadata) {
		root = "archive-pages.zip"
	} else {
		// Find the zip file to use from archive.yml
		// Kept only for backward-compability, will be removed in future versions
		root, err = getArchivePagesRoot(metadata)
		if err != nil {
			return nil, err
		}
	}

	item = &Item{
		Identifier:  identifier,
		LastUpdated: metadata.LastUpdated,
		Root:        root,
		D1:          metadata.D1,
		D2:          metadata.D2,
		Dir:         metadata.Dir,
	}
	SetCacheItem(identifier, item)
	return item, nil
}

func hasArchivePagesZip(metadata *ItemMetadata) bool {
	for _, f := range metadata.Files {
		if f.Name == "archive-pages.zip" {
			return true
		}
	}
	return false
}

func getArchivePagesRoot(metadata *ItemMetadata) (string, error) {

	url := fmt.Sprintf("https://%s%s/archive.yml", metadata.D1, metadata.Dir)
	log.Println(url)
	var pagesMetadata PagesMetadata
	err := readYAML(url, &pagesMetadata)
	//err := Test(pagesMetadata)
	if err != nil {

	}
	return pagesMetadata.Root, nil
}

// GetItemMetadata returns the metadata of an archive.org item.
func GetItemMetadata(identifier string) (*ItemMetadata, error) {
	var metadata ItemMetadata
	url := "https://archive.org/metadata/" + identifier
	err := readJSON(url, &metadata)
	if err != nil {
		log.Fatalf("Failed to fetch metadata: %s", err)
		return nil, err
	}
	return &metadata, nil
}

func readJSON(url string, out interface{}) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)

	err = json.Unmarshal(body, out)
	if err != nil {
		return err
	}
	return nil
}

func readYAML(url string, out interface{}) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)

	log.Printf("%s\n", body)

	err = yaml.Unmarshal(body, out)
	if err != nil {
		return err
	}
	return nil
}

func Test(out interface{}) error {
	body := []byte("version: 1\nitemtype: website\nroot: www.aaronsw.com.zip")
	err := yaml.Unmarshal(body, out)
	if err != nil {
		return err
	}
	return nil
}


cache.go
package main

import (
	"encoding/json"
	"log"
	"os"
	"time"

	"github.com/garyburd/redigo/redis"
)

var pool *redis.Pool

func init() {
	redisURL := os.Getenv("REDIS_URL")
	pool = newPool(redisURL)
}

func GetCacheItem(identifier string) *Item {
	conn := pool.Get()
	defer conn.Close()

	res, err := redis.String(conn.Do("GET", identifier))
	if err == redis.ErrNil {
		return nil
	}
	if err != nil {
		log.Fatalf("Failed to get item from cache - %s: %s", identifier, err)
		return nil
	}
	jsonBody := []byte(res)

	var item Item
	err = json.Unmarshal(jsonBody, &item)
	if err != nil {
		log.Fatalf("Failed to decode the value as json - %s: %s", res, err)
		return nil
	}
	return &item
}

func SetCacheItem(identifier string, item *Item) {
	if item == nil {
		return
	}

	b, err := json.Marshal(item)
	if err != nil {
		log.Fatalf("Failed to encode the value as json - %s: %s", identifier, err)
		return
	}

	conn := pool.Get()
	defer conn.Close()

	_, err = conn.Do("SETEX", identifier, 300, string(b))
	if err != nil {
		log.Fatalf("Failed to set item in redis - %s: %s", identifier, err)
		return
	}
}

func newPool(redisURL string) *redis.Pool {
	return &redis.Pool{
		MaxIdle:     3,
		IdleTimeout: 240 * time.Second,
		Dial: func() (redis.Conn, error) {
			return redis.DialURL(redisURL)
		},
	}
}


archive_pages/webapp.py
from __future__ import annotations
import web
import internetarchive as ia
from jinja2 import Environment, PackageLoader
import io
import yaml
import mimetypes

urls = (
    "/(.*)", "page"
)
app = web.application(urls, globals())
application = app.wsgifunc()

@web.memoize
def get_jinja2_env():
    return Environment(loader=PackageLoader(__name__, "templates"))

def render_template(_filename, **kwargs):
    env = get_jinja2_env()
    template = env.get_template(_filename)
    return template.render(**kwargs)

def get_domain():
    return web.ctx.env.get('HTTP_HOST', '')

def get_itemname():
    domain = get_domain()
    return domain.split(".")[0]

class page:
    def GET(self, path):
        itemname = get_itemname()
        item = self.get_item(itemname)
        return item.render(path)

    def get_item(self, itemname):
        item = Item.find(itemname)
        if not item:
            raise web.notfound("")

        iadata = item.read_archive_yml()
        if iadata and iadata.get("itemtype") == "website":
            root = iadata.get("root") or f"{itemname}.zip"
            return ZipItem(item, root)
        else:
            return item

class ZipItem:
    def __init__(self, item: Item, zip_path: str):
        self.item = item
        self.zip_path = zip_path

    def render(self, path):
        # Internet Archive allows accessing file a/b.txt in file.zip using URL
        # file.zip/a/b.txt
        path = path or "index.html"
        full_path = self.zip_path + "/" + path
        content_type = self.guess_content_type(path)
        if self.is_binary(content_type):
            raise web.redirect(self.item.get_download_url(full_path))
        web.header("Content-type", content_type)
        return self.item.read_file(full_path)

    def is_binary(self, content_type):
        binary_types = ["application", "image"]
        return any(content_type.startswith(t) for t in binary_types)

    def guess_content_type(self, path):
        content_type, charset = mimetypes.guess_type(path)
        if content_type is None:
            content_type = "text/html"
        if content_type.startswith("text/"):
            charset = charset or "utf-8"
            content_type += "; charset=" + charset

        return content_type

CACHE = {}
UNDEFINED = object()

class Item:
    def __init__(self, item: ia.Item):
        self.item = item
        self.files =  [f['name'] for f in item.files]
        self._yamldata = UNDEFINED

    def read_archive_yml(self):
        if self._yamldata is UNDEFINED:
            self._yamldata = self._read_archive_yml()
        return self._yamldata

    def _read_archive_yml(self):
        if "archive.yml" not in self.files:
            return
        contents = self.read_file("archive.yml").decode('utf-8')
        return yaml.safe_load(io.StringIO(contents))

    def get_download_url(self, path):
        return self.item.urls.download + "/" + path

    def get_zipitem(self, path):
        return ZipItem(self, path)

    def has_file(self, path):
        return path in self.files

    def render(self, path):
        if path and not self.has_file(path):
            raise web.notfound("")

        if not path:
            if "index.html" in self.files:
                path = "index.html"
            else:
                return self.render_index()
        return self.read_file(path)

    def render_index(self):
        return render_template(
            "index.html",
            title=self.item.metadata.get('title', ''),
            files=self.item.files)

    def read_file(self, filename) -> bytes:
        file = self.item.get_file(filename)
        response = file.download(return_responses=True)
        return response.content

    @classmethod
    def find(cls, itemname):
        if itemname not in CACHE:
            CACHE[itemname] = cls._find(itemname)
        return CACHE[itemname]

    @classmethod
    def _find(cls, itemname):
        ia_item = ia.get_item(itemname)
        return ia_item and cls(ia_item)


web.go
package main

import (
	"fmt"
	"log"
	"mime"
	"net/http"
	"net/http/httputil"
	"net/url"
	"path/filepath"
	"strings"
)

func handleRequest(w http.ResponseWriter, domain string, path string) {
	log.Printf("REQ: %s %s", domain, path)
	item, err := GetItem(domain)
	if err != nil {
		w.WriteHeader(404)
		w.Header().Set("Content-Type", "text/html")
		fmt.Fprint(w, "<h1>Site Not Found</h1>")
	}

	req, err := http.NewRequest(
		"GET",
		path,
		http.NoBody)

	if err != nil {
		log.Fatalf("Invalid URL")
	}

	proxy := PagesReverseProxy(item)
	proxy.ServeHTTP(w, req)
}

func guessContentType(path string) string {
	ext := filepath.Ext(path)
	if ext == "" {
		return "text/html"
	}
	contentType := mime.TypeByExtension(ext)
	if contentType == "" {
		contentType = "application/octet-stream"
	}
	return contentType
}

func PagesReverseProxy(item *Item) *httputil.ReverseProxy {
	director := func(req *http.Request) {
		path := req.URL.Path
		if strings.HasSuffix(path, "/") {
			path += "index.html"
		}
		path = path[1:] // strip leading /

		req.URL.Scheme = "https"
		req.URL.Host = item.D1

		req.URL.Path = "/view_archive.php"
		req.URL.RawPath = req.URL.Path

		q := url.Values{}
		q.Set("archive", item.Dir+"/"+item.Root)
		q.Set("file", path)
		req.URL.RawQuery = q.Encode()

		if _, ok := req.Header["User-Agent"]; !ok {
			// explicitly disable User-Agent so it's not set to default value
			req.Header.Set("User-Agent", "")
		}
	}
	modifyResponse := func(res *http.Response) error {
		// delete the "Content-Disposition" header as that interfers with opening pages in browser
		delete(res.Header, "Content-Disposition")

		// Set the content-type using the extension of the requested resource
		contentType := guessContentType(res.Request.URL.Query().Get("file"))
		res.Header.Set("Content-Type", contentType)
		res.Header.Set("Cache-Control", "public, max-age=300")
		return nil
	}

	return &httputil.ReverseProxy{
		Director:       director,
		ModifyResponse: modifyResponse,
	}
}

func getItem(domain string) (int, error) {
	// return 0, errors.New("item not found")
	return 0, nil
}

func getSubdomain(host string) string {
	host = strings.Split(host, ":")[0] // strip port
	if host == "onarchive.org" || host == "local.onarchive.org" {
		return "archive-pages"
	}
	return strings.Split(host, ".")[0]
}

// Serves all the archive pages
func Serve() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		handleRequest(w, getSubdomain(r.Host), r.URL.Path)
	})

	fmt.Println("http://localhost:8080/")
	http.ListenAndServe(":8080", nil)
}
